### PR TITLE
Newtype RowIndex(u32)

### LINKED ISSUE #1236: Introduce a newtype for to represent a row in a trace
We should probably introduce a newtype to represent a row. Maybe we call it `Cycle` or `Step`. It could look something like this:
```
pub struct Step(u32);
```

And then we can implement both `From<Step> for u32` and `From<Step> for usize`.

_Originally posted by @bobbinth in https://github.com/0xPolygonMiden/miden-vm/pull/1229#discussion_r1482721654_

### PR DESCRIPTION
**Pull request by sergerad**
Closes #1236.

* Add `struct RowIndex(u32)` in air crate
* Replace `usize` with `Step` for all variables pertaining to execution trace rows
* Implement operator overrides to allow Step to replace `usize`  in arithmetic, indexing, equality, and ordering operations

### CODE REVIEW

**Comment by sergerad** [comment]
@bobbinth if we are happy to do the operator overloading then might be worth using [derive_more](https://docs.rs/derive_more/latest/derive_more/) rather than doing it manually like the PR currently does.

File: processor/src/chiplets/hasher/tests.rs
Comment (sergerad) [inline processor/src/chiplets/hasher/tests.rs:610]: Thoughts on whether we want this to depend on the nightly cfg? If not, should we leave a cleaned up version of this comment about waiting on stable?

Comment (sergerad) [reply processor/src/chiplets/hasher/tests.rs:610]: Same todo comes up a few times

File: air/src/utils.rs
Comment (plafer) [inline air/src/utils.rs:12]: I would move the definition of `Step` to `air/src/trace/row.rs` (module name depends on outcome of previous comment about naming of `Step`)

File: air/src/utils.rs
Comment (plafer) [inline air/src/utils.rs:44]: I think we should use a `value.try_into().expect()` to catch when the `usize` value is too large. The assumption is that we would ensure elsewhere that traces are never longer than `2^32 - 1`, and that we should never hit that panic. But in the case that we'd do, I would prefer to panic over silently wrapping around

File: air/src/utils.rs
Comment (plafer) [inline air/src/utils.rs:26]: This is not guaranteed to work, since the size of `u32` is larger than `i32`. Is there a reason we need to support conversions to `i32` at all?

File: air/src/utils.rs
Comment (plafer) [inline air/src/utils.rs:112]: We should use `rhs.into()` (to use our definition of `usize -> Step` that panics when value doesn't fit in `u32`).

Comment applies everywhere that we use `usize as u32`

**Review by plafer — CHANGES_REQUESTED** [review]
Thank you for taking this on! Left a bunch of mostly minor comments

File: processor/src/chiplets/hasher/tests.rs
Comment (plafer) [reply processor/src/chiplets/hasher/tests.rs:610]: This only comes up in tests, so I think it's fine as is

**Comment by sergerad** [comment]
@plafer I'm part way through addressing your comments. But it has occurred to me that the newtype approach for this type might not yield any benefits beyond readability of the type name given the `u32` type enforces the size invariant. If that is the case, we can get the same benefit through a type alias instead. And we wouldn't have to re-implement all those traits and fns for conversion, arithmetic etc.

Am I missing a reason we would want the newtype over the alias for this specific case? If not, I suggest we update the relevant fns to use this type alias, given the `u32` type fulfills the invariant(s) we care about:
```rust
pub type RowIndex = u32;
```

cc @bobbinth

**Comment by plafer** [comment]
> But it has occurred to me that the newtype approach for this type might not yield any benefits beyond readability of the type name given the u32 type enforces the size invariant. If that is the case, we can get the same benefit through a type alias instead.

The current benefits I see are
- uniform conversion `usize -> u32` (which panics if too large)
- better IDE support (which will should `u32` instead of `Step` when inferring the type)
- Consistent with the current `ContextId` (which is also a newtype)

Main downsides
- more `.into()` calls
- lots of boilerplate in the definition of `Step` with all the required `impl`s

Are there other benefits/downsides that you see? Given the current list, I'd still go for the newtype, if only for the IDE support. The number of `.into()` calls should be quite limited, and (roughly) only happen during serialization/deserialization, or similar. As for the boilerplate definition, we could turn the current boilerplate into a macro, and reuse it for `ContextId`, and any other future types needed.

File: air/src/utils.rs
Comment (sergerad) [reply air/src/utils.rs]: Do we want to panic on overflow for addition and multiplication?

File: air/src/utils.rs
Comment (sergerad) [reply air/src/utils.rs]: I have made these changes except for renaming variables and fns to be "row" instead of "cycle" for now.

File: air/src/constraints/chiplets/bitwise/mod.rs
Comment (sergerad) [inline air/src/constraints/chiplets/bitwise/mod.rs:427]: If I `use crate::RowIndex;` then linter complains of unused import as it is only used in tests.

**Comment by sergerad** [comment]
@plafer latest diffs are quite large. Many of the diffs are due to tests needing to `into()` the `RowIndex`.

I don't think I converted types that shouldn't be a `RowIndex`, but LMK if I did. Thanks!

File: air/src/constraints/chiplets/bitwise/mod.rs
Comment (plafer) [reply air/src/constraints/chiplets/bitwise/mod.rs:427]: Ah, this should be moved in the `air/src/constraints/chiplets/bitwise/tests.rs` file

File: air/src/utils.rs
Comment (plafer) [reply air/src/utils.rs]: I believe we still have some `usize as u32` conversions, e.g. in the `impl Sub<usize> for RowIndex` implementation

File: air/src/utils.rs
Comment (plafer) [reply air/src/utils.rs]: yes I think that's fine, at least for this PR, as this is the current behavior (since we're using primitive types)

**Review by plafer — APPROVED** [review]
Looks great! Approving optimistically - let's just fix the minor nits I left, fix CI (just formatting), and wait for @bobbinth's review as well

File: air/src/trace/rows.rs
Comment (bobbinth) [inline air/src/trace/rows.rs:129]: I'm curious: are there many places where we subtract row index from something?

File: air/src/trace/rows.rs
Comment (bobbinth) [inline air/src/trace/rows.rs:137]: Similar to the above comment: what are some situations when we need to subtract row index from row index? (conceptually, this shouldn't be happening).

File: air/src/trace/rows.rs
Comment (bobbinth) [inline air/src/trace/rows.rs:145]: Similar comment to the ones above. Generally, the way I was thinking about this was that we could add, subtract, etc. values to/from row indexes, but adding two row indexes or subtracting row index from a row index would not be a "native" operation. But maybe there are enough cases to justify this?

File: miden/Cargo.toml
Comment (bobbinth) [inline miden/Cargo.toml:60]: I'm wondering we should re-export `RowIndex` from the `processor` crate. If we do that, this dependency wouldn't be need, I think.

**Review by bobbinth — COMMENTED** [review]
Thank you for doing this! (and sorry that I'm just getting to this PR). This is not a full review yet, but I've left some comments/questions inline.

Also, I'm wondering if these changes impact performance in any noticeable way? I would assume that the complier should erase almost everything related to the extra type conversions - but may be good to double-check.

One way to check this could be to run benchmarks manually. For example, I usually do the following
```bash
make exec-info
MIDEN_LOG=info RAYON_NUM_THREADS=16 ./target/optimized/miden example blake3 -n 100
```
This should run the VM for about 1M cycles. `RAYON_NUM_THREADS` could be set to w/e is best for your machine (I set it to 16 on my Apple M1 Pro chip) and the above benchmark usually takes around 8 seconds.

File: air/src/constraints/chiplets/bitwise/tests.rs
Comment (sergerad) [reply air/src/constraints/chiplets/bitwise/tests.rs:273]: I missed this one. Thanks

File: air/src/options.rs
Comment (sergerad) [reply air/src/options.rs:174]: Agreed. While an inclusive maximum is effectively the same as the index it relates to, it reads strangely because an index is not a count.

If renaming the variables to max_row (or similar) doesn't make sense, then would suggest to use u32 instead.

File: air/src/trace/rows.rs
Comment (sergerad) [reply air/src/trace/rows.rs:129]: Just here
```
let current_shift = NUM_DECOMP_BITS * (OP_CYCLE_LEN - cycle_row - 1);
```

File: air/src/trace/rows.rs
Comment (sergerad) [reply air/src/trace/rows.rs:137]:
```
                // difference between current clock cycle and start clock cycle of the current
                // asmop
                u32::from(a.max(b) - a.min(b.into())) as u8
```

```
                    // compute delta as difference between context IDs, addresses, or clock cycles
                    let delta = if prev_ctx != ctx {
                        (u32::from(ctx) - u32::from(prev_ctx)).into()
                    } else if prev_addr != addr {
                        u64::from(addr - prev_addr)
                    } else {
                        clk - prev_clk - 1
                    };
```

```
impl ChipletsLengths {
    pub fn new(chiplets: &Chiplets) -> Self {
        ChipletsLengths {
            hash_chiplet_len: chiplets.bitwise_start().into(),
            bitwise_chiplet_len: (chiplets.memory_start() - chiplets.bitwise_start()).into(),
            memory_chiplet_len: (chiplets.kernel_rom_start() - chiplets.memory_start()).into(),
            kernel_rom_len: (chiplets.padding_start() - chiplets.kernel_rom_start()).into(),
        }
    }
```

These are all of them.

File: air/src/trace/rows.rs
Comment (sergerad) [reply air/src/trace/rows.rs:145]: This one is unused actually, will remove.

File: air/src/options.rs
Comment (bobbinth) [reply air/src/options.rs:174]: Let's change these back to `u32` values.

File: air/src/trace/rows.rs
Comment (bobbinth) [reply air/src/trace/rows.rs:137]: if these are all of them, I'd probably remove these operator overloading and do manual conversions here.

File: processor/src/chiplets/memory/segment.rs
Comment (sergerad) [reply processor/src/chiplets/memory/segment.rs:74]: @bobbinth just checking whether `clk` here and similar place should be RowIndex or u32 like addr.
```
pub fn read(&mut self, ctx: ContextId, addr: u32, clk: RowIndex)
```

Comment (sergerad) [reply processor/src/chiplets/memory/segment.rs:74]: Also whether this makes sense as RowIndex
```
pub struct MemorySegmentTrace(BTreeMap<RowIndex, Vec<MemorySegmentAccess>>);
```

Comment (bobbinth) [reply processor/src/chiplets/memory/segment.rs:74]: > Also whether this makes sense as RowIndex
> 
> ```
> pub struct MemorySegmentTrace(BTreeMap<RowIndex, Vec<MemorySegmentAccess>>);
> ```

Yeah - I think this should be `pub struct MemorySegmentTrace(BTreeMap<u32, Vec<MemorySegmentAccess>>);` As the key in the map is memory address rather than row index.

File: air/src/trace/rows.rs
Comment (sergerad) [reply air/src/trace/rows.rs]: @bobbinth could you please clarify - do you want all arithmetic overloads removed from RowIndex type?

Currently there is +, -, *, %, shift right, +=, -=.

Thanks.

Comment (bobbinth) [reply air/src/trace/rows.rs]: I think some are fine. For example:

- `RoIndex + u32`, `RoIndex += u32`
- `RoIndex - u32`, `RoIndex -= u32`
- `RoIndex * u32`, `RoIndex *= u32`
- `RoIndex + usize`, `RoIndex += usize`
- `RoIndex - usize`, `RoIndex -= usize`
- `RoIndex * usize`, `RoIndex *= usize`

But things like `RowIndex * RowIndex` or `RowIndex + RowIndex` or bit shifts or modulus operations feel a bit weird.

**Comment by sergerad** [comment]
@bobbinth I ran the benchmarks on M1 also. Seems inconsequential as expected.

next branch:
```
╰─➤  make exec-info                
MIDEN_LOG=info RAYON_NUM_THREADS=16 ./target/optimized/miden example blake3 -n 100
...
--------------------------------
Executed program in 17712 ms
...
--------------------------------
INFO     verify_program [ 1.46ms | 100.00% ]
Execution verified in 1 ms
```

PR branch:
```
╰─➤  make exec-info                
MIDEN_LOG=info RAYON_NUM_THREADS=16 ./target/optimized/miden example blake3 -n 100
...
--------------------------------
Executed program in 16496 ms
...
--------------------------------
INFO     verify_program [ 1.13ms | 100.00% ]
Execution verified in 1 ms
```

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a couple of non-blocking comments inline.

Also, if you don't mind, I'd prefer to merge this after #1419 as I think this PR will be quite a bit easier to rebase than that one.

**Comment by bobbinth** [comment]
@sergerad now that #1419 has been merged, could you rebase your PR? It seems like there is only 1 conflict (for some reasons, I was expecting more).

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

**Pull request merged**
